### DISCUSSION TITLE
The time variable in mesa.Model

### DISCUSSION THREAD

**Discussion by EwoutH** [mesa/mesa#2228] [Future of Mesa]
_This discussion is split of from_
- https://github.com/projectmesa/mesa/pull/2223

## Background
Mesa currently has a `_step` and `_time` counter, added in https://github.com/projectmesa/mesa/pull/1942. In https://github.com/projectmesa/mesa/pull/2223 a lot of discussion has taken place about the usefulness of the `_time` counter.

## Questions for discussion

1. Is a dedicated time/clock system necessary for Mesa, or is the current step counter sufficient?
2. If implemented, should it be a core feature or an optional module?
3. How might this affect the simplicity of creating basic models, especially for beginners?
4. What level of granularity and flexibility should the time/clock system offer? (`Datetime` or not)

**Comment by EwoutH** [comment]
I summarized my points on this subject in these two comments: https://github.com/projectmesa/mesa/pull/2223#issuecomment-2295965337 and https://github.com/projectmesa/mesa/pull/2223#issuecomment-2298453184.

**Comment by Corvince** [comment]
I think time in ABMs is considered in many different senses, but let’s start with the basics. The `step` counter, which represents the number of times `model.step` is called, is universally useful for any model. By convention, we interpret this as a tick between two different model states we are interested in observing. This step counter is crucial for both data collection and visualization purposes. One aspect I particularly appreciate is that it isn't inherently tied to any notion of "time" in the traditional sense. While it shares the characteristic of being linearly increasing in a unidirectional manner, that’s where the similarity to time ends.

Time, in contrast, serves a different purpose. It is often a more abstract concept in ABMs, representing the progression of events or changes in the system rather than just a simple count of iterations. Depending on the model, time can be mapped to real-world intervals, like seconds, minutes, or years, or it can remain an abstract concept used to delineate different phases or states within the simulation. The flexibility of how time is represented and utilized makes this so hard to abstract away. But in any case I think a good solution should not part of a standard model, but something added one. Therefore I would remove the current time counter as a reserved keyword and develop something that can be used additionally.

**Reply by quaquel** [reply]
I agree with this

**Reply by EwoutH** [reply]
I agree with everything up to this point. Very well put (and thanks for joining the conversation)!

> But in any case I think a good solution should not part of a standard model, but something added one. Therefore I would remove the current time counter as a reserved keyword and develop something that can be used additionally.

That's where I differ. I think an _easily overridable_ time counter makes life so much easier. Want to override it? `time=False` and we never touch it again (you practically unreserved it).

I have three very real use-cases:

**1.** model.time can be used for data lookups in time-based lookup tables
```Python
expected_rain = {
    8.00: 3.57,
    8.25: 5.12,
    8.50: 4.89,
    8.75: 6.34,
    9.00: 2.78,
    9.25: 7.65,

class MyModel(Model):
    def __init__(self):
        self.time = 8.00  # Start time at 8:00 AM
    
    def step(self, time=0.25):
        rain = expected_rain[self.time]
        ...
```
Why does this provide value? It help you not to have to convert all your data tables.

**2.** model.time can be used for human-interpretable decision making
```Python
class MyModel(Model):
    def __init__(self):
        self.time = 8.00  # Start time at 8:00 AM
    
    def step(self, time=0.25):
        self.agents.do("something")
        if self.time > 20:
            self.evening_stuff()
```
Why is this useful? You can now change the staring and end time in human readable format. You're not bound by a fixed number of steps from the start. You decouple decision making from the step counter.

**3.** Staged activation can be easier replaced:
```Python
class MyModel(Model):
    def step(self, time=0.25):
        model.agents.do("morning")
        self.time += 0.25
        model.agents.do("afternoon")
        self.time += 0.25
        model.agents.do("evening")
        self.time += 0.25
        model.agents.do("night")
```
And I can go on and on.

Don't want to use any of this stuff? Either:
- Don't touch `model.time`.
- Flip a switch `step(self, time=False)`

**TL;DR:** It's a very convenient built in construct for people that want to do medium-complexity stuff with time. It enables time data lookups, human interpretable time and decision making decoupled from the step counter. It has virtually no downsides since it's very easily overridable and adds extremely low code complexity. It allows steps to be a true tick counter.

**Reply by quaquel** [reply]
Why overwrite it if you can just add your own thing? Having both time and step to me is more confusing then just doing simple step counting only. So, removing time althogether seems simpler and does not prevent the use cases you describe.

**Reply by EwoutH** [reply]
It's a good point, of course you can add any variable and do the same things. However, with model.time:
- An user can really on time always being present
- By _default_, it always increases simultaneously with the step function, even if with another amount
- It saves some lines of code an potential for unintended configuration

All those three things remove complexity and potential failure.

I agree it's an utility function, so it's a nice to have. But it's also a way to allow consistently talk about "real time" (which I think is a useful concepts for real-world ABM).

**Reply by Corvince** [reply]
Your use cases can be achieved simply by doing 

```python
class MyModel(mesa.Model):
    time = 8

    def step():
        self.time += 0.25
        ...
```

I don't think this suffices to be part of mesa. We can show this approach in one of the examples for staged activation. But for beginners its one less variable to know.

**Reply by EwoutH** [reply]
I get your point, but only for beginners it isn't a variable they need to know. Because by default it's not needed. But when you want to modify it, you can.

I think the fact that's always available, interlocks with the step update and the reduced lines of code provide values. It feels more integrated.

**Reply by Corvince** [reply]
Yes but always available and interlocked with step just means beginners may *do* stumble upon it. Either interactively by inspecting model or via the model API documentation or as the parameters for the step function. 

And it saves exactly one line of code, but you still need to add some characters, so it's even less then one line. But I think the biggest downside is that it's part of our public API (for the model.step), so we are also more limited there.

**Reply by quaquel** [reply]
1. Its more then one line of code because it adds a keyword arguments to step. 
2. `Time` might clash with what a user want's to do and thus is a possible source of confusion
3. It creates technological debt. If we want to add a proper clock with datatime as well, we are limited by the fact that time is already there.
4. It clashes with the use of time in the simulator, again a possible source of confusion in users want to go from simple to more advanced use cases.

All this is in my view is too high a price for a simple convenience that @EwoutH sees but other don't really recognize.

**Comment by tpike3** [comment]
For what they are worth my thoughts are that we pause this discussion for now and leave `step` and `time` as is. 

From my experience, I have seen clock time as a fundamental and critical piece of very advanced simulations that were crucial for decision making. 

I think this needs to be a part of Mesa, however, more importantly I think we need to determine what Mesa is trying to be. Please see https://github.com/projectmesa/mesa/discussions/2228 

If this general direction (a robust simulation ecosystem) is what the community adopted, at this time I would say step is the core and then time become an add on library for more advanced users like `cProfile` or `typing` so it creates a scaffolded approach for users, similar to how I see Python.

However, if the community wants to be a beginner-moderate simulation library, then I would say get rid of `time` and focus on accessibility and ease that teaches users core fundamentals of simulations. 

If the community decided some other direction, then that would also influence my recommendation. 

In short, right now, I am recommending a delay, so Mesa can determine what it wants to be and then that can inform our decision.

**Comment by EwoutH** [comment]
I think I have a new perspective that might spark some new discussion.

At some point we had quite some trouble with agent management, since it was abstracted away from the model. You needed to go though the schedulers to get to your agents. But they, different schedulers developed, which stored their agents in also different ways.

What I took from that, it that the Model object is the glue that holds everything together and should - in general - keep the ground truth.

So what I'm afraid of with things like DEVS and `simulator.time`, it that it takes away a ground truth from Mesa again. Because now the Datacollector, Visualisation and such need to detect if there is a simulator or not, and get the time form some place else or not.

Therefor I would propose to keep `model.time` as the ground truth of time in non-strict-ABM models. So:
- `model.steps` strictly keeps what you
- `model.time` keeps a non-discrete time value for all non-strictly-ABM models. If there's a simulator, that simulator updates `model.time`. If not, we can keep it in sync with `model.steps`

This way, you can always trust on `model.time` to be there, if there's a simulator or not. We could let the some components, like the datacollector, choose between using `steps` as index or `time`. By default it could be time (for ABM models), but if you want to collect data on non-fixed intervals you can start using `model.time`. And so on.

**Comment by EwoutH** [comment]
We're having a call on this subject. Note that some of the discussion continues there:
- https://github.com/projectmesa/mesa/discussions/2247

**Comment by EwoutH** [comment]
One challenge I currently have developing new ideas is that we don’t have a consistent way of getting model time.

A solution could be:

1. Adding a `time` property, that gives `model.steps` without a simulator defined or `model.simulator.time` with it.
2. Instantiating a default simulator if no simulator is passed on initialization. That way you can always depend on there being a simulator.

**Reply by quaquel** [reply]
Conceptually, I prefer having a dedicated simulator. However, I am not yet fully satisfied with the current API. The key problem is that the simulator needs a reference to the model, while the model needs a reference to the simulator. What I do like is that the simulator fully encapsulates the run control, which is currently not well-defined in MESA. 

I know some have argued for capturing all information inside the model instance, but, for me, run control is conceptually separate and independent from the model. I thus prefer something like `simulator.run(n=10)` over `model.run(n=10)` (where the latter in turn would do `self.simulator.run(n=10)`.

**Comment by EwoutH** [comment]
This issue popped up in multiple new places. Among others:
- https://github.com/projectmesa/mesa/discussions/2585#discussioncomment-14850500
- https://github.com/projectmesa/mesa/pull/2851

Therefore, I think we should revisit this. I propose we go ahead with this solution:

> * `model.time` keeps a non-discrete time value for all non-strictly-ABM models. If there's a simulator, that simulator updates `model.time`. If not, we can keep it in sync with `model.steps`
> 
> This way, you can always trust on `model.time` to be there, if there's a simulator or not.

**Reply by quaquel** [reply]
I am fine with something along these lines as long as we keep it as minimal as possible to avoid incurring any technological debt.

More broadly, I would favor adding a `RunControl` class. This class would update `model.time`, but also contain `play`, `pause`, support for custom stopping conditions, and support for alternative increments of `model.time` (e.g., 0.25, or full datetime support). Theoretically, we might also make this the entry point for setting replications and seeds. This proposal is based on my experience with various modeling packages, which all have something similar to this.

**Comment by EwoutH** [comment]
This has landed in:
- https://github.com/mesa/mesa/pull/2903
